Limit max batch size for segment allocation, add docs (#13503)
apache/druid: docs/ingestion/tasks.md

  "priority" : 100
}
```
<a name="actions"></a>

## Task actions

These are various overlord actions performed by tasks during their lifecycle. Some typical actions are as follows:
- `lockAcquire`: acquires a time-chunk lock on an interval for the task 
- `lockRelease`: releases a lock acquired by the task on an interval

> - `segmentInsertion`: inserts segments into metadata store

Reviewer: Worth mentioning `segmentTransactionalInsert` too?

- `segmentAllocate`: allocates pending segments to a task to write rows

> - etc.

Reviewer 2: ```suggestion
```

### Batching `segmentAllocate` actions

> In a cluster with a large number of concurrent tasks (say > 1000), `segmentAllocate` actions on the overlord may take very long intervals of time to finish thus causing spikes in the `task/action/run/time`. This may result in lag building up while a task waits for a segment to get allocated.

Reviewer: I think this is going to be useful even at much lower concurrent task counts. I have seen people have issues with allocation timings around task rollover even with just 10s of tasks, requiring a scale-up of metadata store size.

> In a cluster with several concurrent tasks, `segmentAllocate` actions on the overlord may take very long intervals of time to finish thus causing spikes in the `task/action/run/time`. This may result in ingestion lag building up while a task waits for a segment to get allocated.

Reviewer 2: ```suggestion
In a cluster with several concurrent tasks, `segmentAllocate` actions on the overlord can take a long time to finish, causing spikes in the `task/action/run/time`. This can result in ingestion lag building up while a task waits for a segment to be allocated.
```

> The root causes of such spikes are:

Reviewer 2: ```suggestion
The root cause of such spikes is likely to be one of more of the following:
```

- several concurrent tasks trying to allocate segments for the same datasource and interval
- large number of metadata calls made to the segments and pending segments tables 
- concurrency limitations while acquiring a task lock required for allocating a segment

> Since the contention typically arises from tasks allocating segments for the same datasource and interval, the run times can be improved by batching the actions together.

Reviewer 2: ```suggestion
Since the contention typically arises from tasks allocating segments for the same datasource and interval, you can improve the run times by batching the actions together.
```

> Batched segment allocation can be enabled on the overlord by setting `druid.indexer.tasklock.batchSegmentAllocation=true`.See [overlord configuration](../configuration/index.md#overlord-operations) for more details.

Reviewer 2: ```suggestion
To enable batched segment allocation on the overlord, set  `druid.indexer.tasklock.batchSegmentAllocation` to  `true`. See [overlord configuration](../configuration/index.md#overlord-operations) for more details.
```
